PHP 8.1 support

boesing opened the pull request:
|    Q          |   A
|-------------- | ------
| New Feature   | yes

### Description

Initial PHP 8.1 support.

boesing committed:
feature: allow PHP 8.1

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: replace `pdo_sqlite` extension (which does not exist) with `sqlite3`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove expressive related code while bumping coding standard

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: provide proper support for PHP 8.1 in `HtpasswdFactoryTest`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: bump `vimeo/psalm` and get rid of baseline

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented on test/ExceptionTest.php:
```diff
@@ -20,7 +20,8 @@ class ExceptionTest extends TestCase
     public function exception(): iterable
     {
-        $namespace = substr(ExceptionInterface::class, 0, strrpos(ExceptionInterface::class, '\\') + 1);
+        /** @psalm-suppress InvalidLiteralArgument */
```
Looks like a downstream error since the 1st argument is a string and `strrpos` should accept strings.

boesing replied:
https://github.com/vimeo/psalm/issues/6661

boesing committed:
bugfix: ensure `PDOStatement` methods return bool

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

alexraputa commented on src/AuthenticationMiddlewareFactory.php:
```diff
@@ -6,21 +6,16 @@
 
 class AuthenticationMiddlewareFactory
 {
     public function __invoke(ContainerInterface $container): AuthenticationMiddleware
```
We can simplify it.

```suggestion
    public function __invoke(ContainerInterface $container): AuthenticationMiddleware
    {
        $authentication = $container->get(AuthenticationInterface::class);

        if (null === $authentication) {
            throw new Exception\InvalidConfigException(
                'AuthenticationInterface service is missing'
            );
        }

        Assert::isInstanceOf($authentication, AuthenticationInterface::class);

        return new AuthenticationMiddleware($authentication);
    }
```

boesing replied:
Actually that wont work because fetching non-existing services using `ContainerInterface#get` will throw `ContainerExceptionInterface`. 

So we have to check if the service is available and then let assert do its magic to ensure old behavior.

boesing replied:
https://github.com/php-fig/container/blob/2ae37329ee82f91efadc282cc2d527fd6065a5ef/src/ContainerInterface.php#L17

weierophinney committed:
feat: remove support for PHP 7.3

- Bumps minimum supported PHP version to 7.4
- Bumps PHPUnit to 9.5.5 (PHP 8.1 support)
- Bumps Propecy to 1.14 (PHP 8.1 support)

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: treat deprecations as exceptions when testing

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
